### ISSUE TITLE
Gaze angles are incorrect for Fish eye lens

### ISSUE THREAD

**Issue by DayanandVH** [TadasBaltrusaitis/OpenFace#347]
Hi Tadas,

Congratulations on the OpenFace project. Its really great work as so many features are included in it. 

I'm planning to use OpenFace for one of my project. But I have one issue with it. I'm attaching fisheye lens to my webcam and collecting videos. When I extracted features(I'm mainly interested in Gaze angles) of videos collected using fisheye lens using FeatureExtraction command the output is totally wrong. Even tough I'm looking at the all the corners of my laptop screen in the video(fisheye) the output shows almost same gaze angles throughout the video with slight variations. I undistorted the collected fisheye videos to normal lens videos. OpenFace output for converted video also showing similar wrong output(it was showing same Gaze angles throughout the video).   
But when tested same command for same webcam video which is collected without fisheye lens(or by removing fisheye lens), the output is showing almost correct gaze angles. Is there any problem with OpenFace when using with fisheye lens or do I need to make any modifications to make OpenFace to work with Fisheye videos. Any suggestions or help would be appreciated.
Thanks

**Comment by spamrefuse** [comment]
I have no answer to your problem, but we use a Samsung Gear 360 degrees camera, which has kind of two fisheye lenses. I have a hunch that the default values for the focal lengths and/or optical centers are incorrect and need to be set on the command line. However, I'm still puzzled what the values might be for a fisheye lens.....

**Comment by TadasBaltrusaitis** [comment]
I have never tested OpenFace on a fisheye or other very wide angle lenses so do not know how well it would work, but I'm not that surprised that it fails. 

How bad is the distortion? I would have suggested that you undistort the fisheye lens, but it seems you have already tried that. Have you calibrated the camera with the fisheye lens? What were the camera intrinsic parameters?

**Comment by DayanandVH** [comment]
@TadasBaltrusaitis 
Yes.. I calibrated the camera with fisheye lens and the camera intrinsic parameters are:  fx=325.7443031474828 fy=324.04749716943536 cx=324.6947804521985 cy=247.38459525552037.
( Distortion Coefficient: -0.1641841110559771, 0.4624751347191517, -0.6720374285731858, 0.2854720910461931) 
I'm attaching these two graphs of Gaze angles (Yaw in blue and Pitch in Saffron) for the videos which I captured without attaching fisheye and undistorted fisheye videos. During the video I'm looking at center and all corners of my laptop screen for 10 sec each.
![gazeangle_without](https://user-images.githubusercontent.com/32278839/36600324-398943a4-18b2-11e8-8a49-db7144c0e450.png)
![gazeangle_calibrated](https://user-images.githubusercontent.com/32278839/36600338-4116da82-18b2-11e8-92e2-e903d4a6507a.png)
As I mentioned in my last post Openface is showing wrong angles for undistorted videos but it shows nearly correct angles for videos collected without attaching fisheye lens. 
I'm surprised as undistorted video quality is same as videos collected without fisheye lens and even in undistorted videos I can see pupil movement (in different directions) clearly but Openface is showing same Gaze angle (mostly showing as Gaze position at center of my laptop screen) throughout the video as shown in graph. I'm wondering why Openface is not able to recognize change in Gaze direction in fisheye lens undistorted videos while it does a good job with normal videos (without fisheye lens).
I hope this information will help you to provide me some suggestions to solve this problem.

Thanks

**Comment by TadasBaltrusaitis** [comment]
That's interesting, I wonder if this is to do with the calibration or computation of the angle or with difficulty of tracking an iris in a fisheye lens image. Is OpenFace still able to track the iris in the fisheye images?

Another way to debug this would be to see if the head pose is still correct in such videos. Have you looked at the head pose angles? This might be easier to debug as head pose is easier to track and is generally more reliable.

**Comment by DayanandVH** [comment]
@TadasBaltrusaitis , Openface successfully recognized iris in fisheye and calibrated fisheye videos. Also I checked head pose angles for both normal and calibrated videos and no issues exists there as no such difference between head pose angles of two videos and it looks correct to me.

**Comment by TadasBaltrusaitis** [comment]
I was thinking about it a bit and I believe that fish-eye lenses would require modification to the mathematics of how the gaze and head pose are worked out, as a lot of the assumptions that are used  (a pinhole camera model) tend to break down in case of fish-eye lenses. This does sound like an interesting project though!

**Comment by jsf2167** [comment]
Hi @TadasBaltrusaitis,

Thanks for this project and your willingness to stay opensourced. OpenFace has been an increible boon to our research in brain dynamics between parents and children. Due to our fMRI context, the camera is very close to the face and we use a fisheye lens to maximize the face that can be captured at short distances (between the face and the camera). 

From our experience, the AUs seem to match reality, breaking down only in limited cases of obstruction of the face, awkward pose angles, and/or racial/ethinc minority faces. However, getting any gaze or pose information appears completely distorted, primarily in the Z dimension (and all calculations relying on it). 

I am curious, beyond attempting camera calibration as suggested across multiple threads here, do you have any suggestions for working with video captured from fisheye lenses, or plans to add functionality to OpenFace to resolve it?

Thanks again for all of your work here, as I mentioned, it has been incredibly informative to our work, and opened up vast new directions for it.
